File: 016-retirement_py/retirement.py
def update_stage(age, saving, rate, change):
    """
    update stage for age and savings
    """
    print(f"Age {age // 12:3d} month {age % 12:2d} you have ${saving:,.2f}")
    with open("ans.txt", 'a') as f:
        f.write(f"Age {age // 12:3d} month {age % 12:2d} you have ${saving:,.2f}\n")
    age += 1
    saving = (1 + rate / (12 * 100)) * saving + change
    return age, saving


def retirement(start_age, initial_savings, working_info, retired_info):
    """
    Prints the current status of an individual's retirement account.
    The dictionaries both have these fields: 
      "months","contribution","rate_of_return"

    Args:
      start_age (int): At what age (in months) does the individual start
      initial_savings (float): initial savings in dollars
      working_info (dict): information about working
      retired_into (dict): information about retirement

    Returns:
    None
    """
    age = start_age
    saving = initial_savings
    for m in range(WORKING_INFO["Months"]):
        age, saving = update_stage(age, saving, WORKING_INFO["Rate of Return"], WORKING_INFO["Per Month Savings"])
    for m in range(RETIRED_INFO["Months"]):
        age, saving = update_stage(age, saving, RETIRED_INFO["Rate of Return"], RETIRED_INFO["Per Month Spendings"])


WORKING_INFO={"Months": 489, "Per Month Savings": 1000, "Rate of Return": 4.5}
RETIRED_INFO={"Months": 384, "Per Month Spendings": -4000, "Rate of Return": 1}
retirement(327, 21345, WORKING_INFO, RETIRED_INFO)

File: 037-retirement_cpp/retirement.cpp
#include <iostream>
#include <fstream>
#include <iomanip>  // For formatting the output
using namespace std;

class RetireInfo {
    public:
    int months;
    double contribution;
    double rate_of_return;
};


pair<int, double> update_stage(int age, double saving, double ror, double contribution){
    cout << "Age " << setw(3) << age / 12 << " month " << setw(2) << age % 12 << " you have $" << fixed << setprecision(2) << saving << endl;
    
    ofstream outfile;
    outfile.open("ans.txt", ios_base::app);
    outfile << "Age " << setw(3) << age / 12 << " month " << setw(2) << age % 12 << " you have $" << fixed << setprecision(2) << saving << endl;
    outfile.close();
    
    age += 1;
    saving = (1 + ror / (12 * 100)) * saving + contribution;
    return make_pair(age, saving);
}


/**
* @brief Calculates and prints the current balance as an individual works and retires
*
* @param startAge The time in months when the model beings
* @param initial The initial amount of savings an individual has
* @param working Details for the model about how long an individual  works with savings and investment gains
* @param retired Details for the model about how long the individual lives in retirement with expenditures and continued investment gains
*/
void retirement (int startAge, double initial, RetireInfo working, RetireInfo retired){
    int age = startAge;
    double saving = initial;

    for (int m = 0; m < working.months; ++m){
        tie(age, saving) = update_stage(age, saving, working.rate_of_return, working.contribution);
    }
    for (int m = 0; m < retired.months; ++m){
        tie(age, saving) = update_stage(age, saving, retired.rate_of_return, retired.contribution);
    }
}

int main(){
    RetireInfo working, retired;
    working.months = 489;
    working.contribution = 1000;
    working.rate_of_return = 4.5;
    retired.months = 384;
    retired.contribution = -4000;
    retired.rate_of_return = 1;

    retirement(327, 21345, working, retired);

    return 0;
}